### ISSUE TITLE
More distribution to help with bottleneck operations (e.g. shuffling)

### ISSUE THREAD

**Issue by wctgit** [gcp/leela-zero#783]
This is a general suggestion, but it has some specific implications for the immediate situation.

Currently, as far as I have gathered based on recent posts/comments, it seems that training has hit a new bottleneck with the shuffling operation, being dependent on hard drive operations. I don't know 100% for sure that this suggestion will help with that bottleneck, but I have a hunch that it might (let's say roughly 75% estimated chance of helping).

Many clients, such as my own machine, have an excess of additional resources beyond simply GPUs. I have a fairly large amount of RAM available (for a basic laptop), and a fairly large amount of disk space available. If my machine could help with any bottlenecks in those areas, then of course I'd be more than willing to lend the resources.

For example, if the shuffling of games (and presumably board positions, I'm guessing) could be distributed among many clients in parallel, with the results sent back in aggregate to the server -- or even if the results were sent back on an as-needed basis -- then perhaps this could alleviate the current bottleneck of shuffling games/boards on one machine. It seems to me intuitively obvious that such shuffling is an inherently parallelizable operation.

Games themselves are extremely small in terms of storage size, so it's entirely feasible that clients could store a large number of games locally. Perhaps many clients could literally store all the games locally for the cost of a mere few GB of storage, uncompressed. This massive distribution of redundant data would allow the server to merely send IDs (or hashes) of the games to be shuffled and/or from which to generate random/shuffled board positions, rather than having to send the entire game (or batch of games) to the client for each shuffle request. This would allow shuffling to act more like a distributed database operation than a local disk-intensive operation.

Indeed, just as google does most of its work on RAM, rather than disk, perhaps each client could have a large chunk of a few GB of RAM dedicated to a cache of shuffled games and/or board positions. Personally, I have 16 GB of RAM, and rarely use more than half of that. I could easily donate 4 or even up to 8 GB of RAM for caching/shuffling purposes. And I've only got a run-of-the-mill laptop. I bet most/many of the existing clients have even more they could donate.

Anyway, the example of the shuffling operation is just one example of how additional bottleneck operations could be done more distributively.

What would be needed is some configuration options for AutoGTP to allow people to specify how much RAM and HDD space they have available to donate (perhaps with some reasonable defaults which can be estimated based on the device being used). Also, update AutoGTP and/or the server with which it communicates to generate new kinds of distributed operations and to perform them. It's hard for me to estimated how difficult or how long it would take to make such changes (the technology has changed so much recently and I'm a bit out of date), but considering what AutoGTP and the server already accomplish, I can't imagine it's outside the realm of reasonableness. :-)

As the project develops, the server admins can decide how much resources are needed to overcome whatever bottlenecks are encountered, and allocate resources and jobs to various client machines to alleviate said bottlenecks. Perhaps some machines have tons of RAM but little HD space; perhaps others it's the opposite. Some machines will be slow in operation but have a fast network connection; others the opposite. Like Bittorrent and other distributed networks, perhaps clients could even coordinate some of the distribution efforts amongst themselves. For example, instead of all the clients downloading games directly from the server, they could form a torrent-like swarm and download the games from other clients, spreading the network demands around rather than dumping all the responsibility on the server. (Even shuffling could presumably be done this way.) Etc.

**Comment by marcocalignano** [comment]
I even have a better suggestion. instead of download process and upload the data, at the end of each game, autogtp could already prepare the chunks, shuffle them compresse them and send them to the server database together with the game. in this way each client has less to work (no moving data around the internet) and can prepare the chunk for training already. Then when @gcp check if the Game is legit the training process can take the ready chunk from the db shuffle one more time and use them.

**Comment by jkiliani** [comment]
I would be very surprised if this helped. As far as I understand, the current shuffling happens when the chunks are parsed and puts considerable strain on the combination of CPU and RAM of the training machine. Pre-shuffling, i.e. shuffling already when making the chunks, makes the file compression of the chunks inefficient, which results in much larger file sizes of the training data. Removing all order from the chunks presumably also makes debugging any problems with training problematic.

**Comment by gcp** [comment]
>it seems that training has hit a new bottleneck with the shuffling operation, being dependent on hard drive operations.

No, it's not.

This is just a bad idea for a whole bunch of reasons, and I'm sure many of those will become extremely obvious if you just give this a bit more thought.

>Games themselves are extremely small in terms of storage size, so it's entirely feasible that clients could store a large number of games locally. Perhaps many clients could literally store all the games locally for the cost of a mere few GB of storage, uncompressed. 

As was explained in the thread discussing the *actual* problem, the dataset is around 23G compress *without shuffling* and ~230G uncompressed.

**Comment by wctgit** [comment]
Again, as I said, this issue is a general suggestion, not specific to the shuffling issue. There are other operations which could be distributed. Please consider re-opening the issue with that possibility in mind.

Besides, 

1. To many people, 230GB is still a reasonable amount of data storage they could donate. Nearly all run-of-the-mill laptops have at least/over 1TB of HD space these days, and a lot of that goes unused. 
2. Many people have fast and unlimited internet connections (myself included) and downloading 230GB is actually not unreasonable for them. 
3. Certainly downloading 23 GB is reasonable on most/many machines and internet connections. 
4. Downloading can be distributed amongst clients a la torrents, meaning that the length of time it takes **all** of the distributed clients to download the data is on the order of roughly O(1) from the server's perspective; the server, if internet speed is a constraint, could theoretically just upload the data **once** and the faster clients would quickly distribute it amongst themselves and the other clients. This is routine on torrent networks. Microsoft is also now using this model to speed up deployment of Windows updates.
5. Once clients have done the initial download, they would only need to download relatively small future updates to the training data, and again this can be done distributively. It's not like they would have to download 23GB per day, right? There aren't that many games generated per day, only on the order of 100,000, which at less than 10KB each is on the order of 1GB uncompressed, or 100MB compressed. This is easily manageable on many/most run-of-the-mill clients.
6. Each client doesn't even need to download the entire dataset anyway; that was just an example showing that it's not infeasible. You could still generate a lot of parallelization by having clients download only relatively small portions of the data, much like how a RAID system works. 500 clients collectively sharing 100 redundant copies of the dataset would only require storing 230*100/500 = 46 GB uncompressed, or 4.6 GB compressed per machine. This is easily feasible, and also routine among torrent networks where many clients don't have the full download (i.e. seed), but there is enough redundancy within them to reproduce the full download 100 times over.
7. Again, the shuffling example was merely an example. I don't think this suggestion overall is 'invalid' even if I'm way-off-base regarding the shuffling example. The main point of this issue is that there are loads and loads of additional resources that clients have available that could be utilized to help offload bottle-necky operations from the server. 

Given all that, however, I will read the thread dealing with the actual shuffling issue more carefully to see what the actual underlying issues are. However, I'm at a loss as to which thread exactly is the main one. I recall reading about it scattered in various threads. Could you point me to the correct main thread?

**Comment by jkiliani** [comment]
I think #747, #757 are most specific to the issue.

**Comment by wctgit** [comment]
thx

**Comment by Ttl** [comment]
What's the point of distributing the training data to clients? Storage is not an issue. Server can already handle it fine.

**Comment by wctgit** [comment]
The shuffling. Shuffling is essentially a sort operation with the sort key uniformly randomized. Sorting is an inherently distributable task.

Again, perhaps shuffling isn't the best example. Perhaps I'm overlooking a key constraint that needs to occur on one single machine to be efficient. But some of the objections were in regards to storage size, so I'm just trying to show a 'worst case scenario to best case scenario' that even if each client was required to store the entire training data set, it's not unreasonable or infeasible to do so on most people's commodity-level hardware.

**Comment by Ttl** [comment]
Server can do the shuffling by itself when writing the chunks if needed. It's incredibly wasteful to do it on hundreds of clients when one computer is enough to do it.

**Comment by wctgit** [comment]
But isn't it currently a bottleneck? Sometimes using resources wastefully can be worthwhile to overcome a bottleneck. Caches are a good example. Google does nearly all of its search operations in RAM. Compared to HDDs, RAM is incredibly expensive per GB, but it's still worth it to Google to do so.

**Comment by Ttl** [comment]
Reading in random order from compressed chunks that are written in order is the problem. It's trivial to just write the chunks in shuffled order in the first place.

If I'm understanding your proposal correctly you would have the server send the training data to clients over network, clients would reorder the data and send it back. It would be faster for server to just do the reordering by itself.

**Comment by marcocalignano** [comment]
Is it really so expensive to let autogtp do the job below for just one game then, pack the data and send it to the server db?
```
for chunk in chunks:
                with gzip.open(chunk, 'r') as chunk_file:
                    file_content = chunk_file.readlines()
                    item_count = len(file_content) // DATA_ITEM_LINES
                    # Pick only 1 in every 8 positions
                    picked_items = random.sample(range(item_count),
                                                 (item_count + 7) // 8)
                    for item_idx in picked_items:
                        pick_offset = item_idx * DATA_ITEM_LINES
                        item = file_content[pick_offset:pick_offset + DATA_ITEM_LINES]
                        str_items = [str(line, 'ascii') for line in item]
                        # Pick a random symmetry to apply
                        symmetry = random.randrange(8)
                        success, data = self.convert_train_data(str_items, symmetry)
```

**Comment by Ttl** [comment]
Issue in training that requires shuffling is that there are too many positions from the same game in the buffer at the same time. Reordering moves in a single game doesn't help with that.

**Comment by marcocalignano** [comment]
you could maybe use the ```linereader``` module to read only the line you need.
```
from linereader import dopen
file = dopen('C:/Python34/README.txt')

file.seekline(58)
line_58 = file.readline()
next_10_lines = file.readnext(10)
line_67 = file.getline(67)
```

**Comment by wctgit** [comment]
> Reading in random order from compressed chunks that are written in order is the problem. It's trivial to just write the chunks in shuffled order in the first place.

Doesn't this assume a single shuffle order? Shouldn't there be an independent shuffle between some number of training sessions? Again, it could lead to overfitting if only a single shuffle is used.

> you would have the server send the training data to clients over network, clients would reorder the data and send it back.

Well, in my example a) the clients could distribute the downloading of data between themselves, so that the server would not have to repeatedly send the training data, b) the server would only have to send a shuffle job request, perhaps with some data IDs associated with it, or other limited parameters if necessary; IDs and parameters are a lot smaller than the full data.
